src/machine/config.rs

```diff
@@ -1,32 +1,191 @@
-pub struct MachineConfig {
-    pub streams: StreamConfig,
-    pub toplevel: &'static str,
+use rand::{rngs::StdRng, SeedableRng};
+
+use crate::Machine;
+
+use super::{
+    bootstrapping_compile, current_dir, import_builtin_impls, libraries, load_module, Atom,
+    CompilationTarget, IndexStore, ListingSource, MachineArgs, MachineState, Stream, StreamOptions,
+};
+
+/// Describes how the streams of a [`Machine`](crate::Machine) will be handled.
+pub struct StreamConfig {
+    inner: StreamConfigInner,
+}
+
+impl Default for StreamConfig {
+    /// Defaults to using in-memory streams.
+    fn default() -> Self {
+        Self::in_memory()
+    }
+}
+
+impl StreamConfig {
+    /// Binds the input, output and error streams to stdin, stdout and stderr.
+    pub fn stdio() -> Self {
+        StreamConfig {
+            inner: StreamConfigInner::Stdio,
+        }
+    }
+
+    /// Binds the output stream to a memory buffer, and the error stream to stderr.
+    ///
+    /// The input stream is ignored.
+    pub fn in_memory() -> Self {
+        StreamConfig {
+            inner: StreamConfigInner::Memory,
+        }
+    }
 }
 
-pub enum StreamConfig {
+enum StreamConfigInner {
     Stdio,
     Memory,
 }
 
-impl Default for MachineConfig {
+/// Describes how a [`Machine`](crate::Machine) will be configured.
+pub struct MachineBuilder {
+    pub(crate) streams: StreamConfig,
+    pub(crate) toplevel: &'static str,
```

Comment:
Do we want to require the top level to be a `&'static str`?

I could see it being useful e.g. to use a file loaded at runtime to be used,
in that case we could use a `Cow<'static, str>` to allow either a `&'static str` or a `String` to be used.

We would need to adjust `MachineBuilder::with_toplevel` and `Machine::load_top_level` accordingly. 

`Machine::load_top_level` would need to construct `toplevel_stream` depending on the Cow variant we have either via `Stream::from_static_string` or `Stream::from_owned_string`. 

Maybe a `Stream::from_string` taking a `Cow<&'static, str>` would make sense? (name is debatable other suggestions `from_cow_string` or `from_static_or_owned_string`)

Reply:
I mentioned this above actually. I decided not going this way because it seemed like it would need really deep changes to accommodate properly. Maybe just leaking for now if it is too complicated is a good idea to get the interface right faster.

Reply:
I would have expected this to rather straight forward, but it's not unlike that I missed something which makes this more difficult than immediately obvious.

I see 4 options here:
- keep the interface as is for now, making a breaking change later when changing to cow
- keep the interface as is, later add another e.g. with_owned_top_level taking a cow or string, the current method can then just wrap the reference into a cow
- rename the function to e.g. with_static_top_level, so we can add with_owned_top_level later (mirror the functions on Stream)
- change to a cow now and leak in the owned case

Reply:
It's much more likely that I overestimated how much work it would actually be. If it's as easy as you seem to imply I'll probably do it in this PR, although I'm not sure if `Cow<'static, str>` is the best type to put in the API here (it probably is, but I want to think about it a bit more).

Reply:
As far as I can see the top_level string only ends up being used to create the `toplevel_stream` in [`Machine::load_top_level`](https://github.com/mthom/scryer-prolog/pull/2582/files#diff-327437ea05a8167d7cbbff2a15c291a3da408980aa9144ab5d471f4fff9d51abL307).

Unless something unbeknownst to me depends on this being a Stream created from a static string I would assume that all that needs to be changed is the plumbing (i.e. argument type to `Machine::load_top_level` and type of the `toplevel` field in `MachineBuilder`) and how the `toplevel_stream` is constructed i.e.

```diff
-         let toplevel_stream = Stream::from_static_string(program, &mut self.machine_st.arena);
+         let toplevel_stream = match program {
+             Cow::Borrowed(program) => Stream::from_static_string(program, &mut self.machine_st.arena),
+             Cow::Owned(program) => Stream::from_owned_string(program, &mut self.machine_st.arena)
+         };
```

Reply:
Maybe rather than having Cow in the API it would make sense to have `with_static_top_level` taking `&'static str` and `with_owned_top_level` taking `String` rather than having a single method that takes `Cow` or it could be a single method taking an `impl Into<Cow<'static, str>>`?